File: src/main/java/com/cardgames/server/game/GameState.java
package com.cardgames.server.game;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.List;

/**
 * Server-side Klondike game state.
 *
 * Mirrors the JS buildGameState() / applyDrawStock() / move logic exactly
 * so we can replay the client's move history and validate it server-side
 * (DEV-59).
 *
 * Move token format (matches MOVE_TYPES in gameEngine.js):
 *   draw          - draw up to drawCount cards from stock (or flip waste)
 *   wt:col        - top waste card → tableau column col (0-based)
 *   wf            - top waste card → foundation (suit determines pile)
 *   tt:from:to    - card stack from tableau col → tableau col
 *   tf:from       - top of tableau col → foundation
 *   abandon       - game abandoned (no further validation)
 */
public class GameState {

    private static final String[] SUIT_ORDER = { "clubs", "diamonds", "hearts", "spades" };

    private final int drawCount;

    // Tableau: 7 piles
    private final List<List<Card>> tableau = new ArrayList<>();
    // Foundations: indexed by suit order above (0=clubs 1=diamonds 2=hearts 3=spades)
    private final List<List<Card>> foundations = new ArrayList<>();
    // Stock and waste as stacks (front = top)
    private final Deque<Card> stock = new ArrayDeque<>();
    private final List<Card>  waste = new ArrayList<>();

    private int moveCount = 0;

    // ── Build initial state from a shuffle seed (DEV-73) ──────────────────
    public GameState(long seed) {
        this(seed, "draw3");
    }

    public GameState(long seed, String drawMode) {
        this(SeededShuffle.shuffle(seed), drawMode);
    }

    // ── Build initial state from 52 ordered card IDs ──────────────────────
    public GameState(int[] cardIds) {
        this(cardIds, "draw3");
    }

    public GameState(int[] cardIds, String drawMode) {
        this.drawCount = "draw1".equals(drawMode) ? 1 : 3;
        // Deal tableau: pile i has i+1 cards, last is face-up
        int idx = 0;
        for (int i = 0; i < 7; i++) {
            List<Card> pile = new ArrayList<>();
            for (int j = 0; j <= i; j++) {
                pile.add(new Card(cardIds[idx], j == i));
                idx++;
            }
            tableau.add(pile);
        }
        // Remaining 24 cards go to stock face-down (first card = top of stock)
        for (int i = idx; i < 52; i++) {
            stock.push(new Card(cardIds[i], false));
        }
        // Reverse so stock[0] is drawn first (push inverts order)
        Deque<Card> ordered = new ArrayDeque<>();
        for (Card c : stock) ordered.addFirst(c);
        stock.clear();
        stock.addAll(ordered);
        // Foundations
        for (int i = 0; i < 4; i++) foundations.add(new ArrayList<>());
    }

    // ── Replay a comma-separated turns string ─────────────────────────────
    public ReplayResult replay(String turns) {
        if (turns == null || turns.isBlank()) {
            return new ReplayResult(false, "Empty turns string", moveCount);
        }
        String[] tokens = turns.split(",");
        for (String raw : tokens) {
            String token = raw.trim();
            if (token.isEmpty()) continue;
            if ("abandon".equals(token)) break; // abandoned — stop here
            boolean ok = applyMove(token);
            if (!ok) {
                return new ReplayResult(false, "Invalid move: [" + token + "] after " + moveCount + " moves", moveCount);
            }
        }
        return new ReplayResult(true, "OK", moveCount);
    }

    // ── Apply a single move token ──────────────────────────────────────────
    private boolean applyMove(String token) {
        String[] parts = token.split(":");
        switch (parts[0]) {
            case "draw": return applyDraw();
            case "wt":   return parts.length >= 2 && applyWasteToTableau(Integer.parseInt(parts[1]));
            case "wf":   return applyWasteToFoundation();
            case "tt":   return parts.length >= 3 && applyTableauToTableau(Integer.parseInt(parts[1]), Integer.parseInt(parts[2]));
            case "tf":   return parts.length >= 2 && applyTableauToFoundation(Integer.parseInt(parts[1]));
            default:
                return false;
        }
    }

    // ── Draw (DEV-15, DEV-58) ─────────────────────────────────────────────
    private boolean applyDraw() {
        if (stock.isEmpty()) {
            // Flip waste back to stock
            for (int i = waste.size() - 1; i >= 0; i--) {
                Card c = new Card(waste.get(i));
                c.setFaceUp(false);
                stock.addFirst(c);
            }
            waste.clear();
        } else {
            int count = Math.min(drawCount, stock.size());
            for (int i = 0; i < count; i++) {
                Card c = stock.poll();
                c.setFaceUp(true);
                waste.add(c);
            }
        }
        moveCount++;
        return true;
    }

    // ── Waste → Tableau ───────────────────────────────────────────────────
    private boolean applyWasteToTableau(int col) {
        if (waste.isEmpty() || col < 0 || col >= 7) return false;
        Card card = waste.get(waste.size() - 1);
        List<Card> pile = tableau.get(col);
        if (!canPlaceOnTableau(card, pile)) return false;
        waste.remove(waste.size() - 1);
        tableau.get(col).add(card);
        moveCount++;
        return true;
    }

    // ── Waste → Foundation ────────────────────────────────────────────────
    private boolean applyWasteToFoundation() {
        if (waste.isEmpty()) return false;
        Card card = waste.get(waste.size() - 1);
        int fIdx = suitIndex(card.getSuit());
        if (fIdx < 0) return false;
        List<Card> foundation = foundations.get(fIdx);
        if (!canPlaceOnFoundation(card, foundation)) return false;
        waste.remove(waste.size() - 1);
        card.setFaceUp(true);
        foundation.add(card);
        moveCount++;
        return true;
    }

    // ── Tableau → Tableau (moves entire face-up stack) ────────────────────
    private boolean applyTableauToTableau(int from, int to) {
        if (from < 0 || from >= 7 || to < 0 || to >= 7 || from == to) return false;
        List<Card> srcPile  = tableau.get(from);
        List<Card> destPile = tableau.get(to);

        // Find the first face-up card (bottom of the moving stack)
        int splitIdx = -1;
        for (int i = 0; i < srcPile.size(); i++) {
            if (srcPile.get(i).isFaceUp()) { splitIdx = i; break; }
        }
        if (splitIdx < 0) return false;

        Card bottomOfStack = srcPile.get(splitIdx);
        if (!canPlaceOnTableau(bottomOfStack, destPile)) return false;

        List<Card> moving = new ArrayList<>(srcPile.subList(splitIdx, srcPile.size()));
        srcPile.subList(splitIdx, srcPile.size()).clear();

        // Flip new top of source pile
        if (!srcPile.isEmpty()) srcPile.get(srcPile.size() - 1).setFaceUp(true);

        destPile.addAll(moving);
        moveCount++;
        return true;
    }

    // ── Tableau → Foundation ──────────────────────────────────────────────
    private boolean applyTableauToFoundation(int col) {
        if (col < 0 || col >= 7) return false;
        List<Card> srcPile = tableau.get(col);
        if (srcPile.isEmpty()) return false;
        Card card = srcPile.get(srcPile.size() - 1);
        if (!card.isFaceUp()) return false;
        int fIdx = suitIndex(card.getSuit());
        if (fIdx < 0) return false;
        List<Card> foundation = foundations.get(fIdx);
        if (!canPlaceOnFoundation(card, foundation)) return false;
        srcPile.remove(srcPile.size() - 1);
        if (!srcPile.isEmpty()) srcPile.get(srcPile.size() - 1).setFaceUp(true);
        card.setFaceUp(true);
        foundation.add(card);
        moveCount++;
        return true;
    }

    // ── Validation rules ──────────────────────────────────────────────────
    private boolean canPlaceOnTableau(Card card, List<Card> pile) {
        if (pile.isEmpty()) return card.getRank() == 13; // King only
        Card top = pile.get(pile.size() - 1);
        if (!top.isFaceUp()) return false;
        return top.getRank() == card.getRank() + 1 && top.isRed() != card.isRed();
    }

    private boolean canPlaceOnFoundation(Card card, List<Card> foundation) {
        if (foundation.isEmpty()) return card.getRank() == 1; // Ace only
        Card top = foundation.get(foundation.size() - 1);
        return top.getSuit().equals(card.getSuit()) && top.getRank() == card.getRank() - 1;
    }

    // ── Win check ─────────────────────────────────────────────────────────
    public boolean isWon() {
        return foundations.stream().allMatch(f -> f.size() == 13);
    }

    // ── Helpers ───────────────────────────────────────────────────────────
    private int suitIndex(String suit) {
        for (int i = 0; i < SUIT_ORDER.length; i++) {
            if (SUIT_ORDER[i].equals(suit)) return i;
        }
        return -1;
    }

    public int getMoveCount() { return moveCount; }
}

File: ios-app/KlondikePro/Engine/GameState.swift
import Foundation

/// Full Klondike solitaire game state.
///
/// This is a pure value type — mutations return a new `GameState` (or mutate in-place
/// depending on call site). The `GameStore` `@Observable` wraps this and publishes changes.
///
/// Move recording: every legal move appends a token to `turns` in the same format
/// accepted by the Spring Boot replay validator:
///   draw, wt:col, wf, tt:fromCol:fromIdx:toCol, tf:col, ft:foundationIdx:toCol
struct GameState {

    // MARK: - Board state

    var stock:     [Card]         // remaining draw pile, face down
    var waste:     [Card]         // drawn cards, face up (top = last)
    var tableau:   [[Card]]       // 7 columns; top of each = last element
    var foundation: [Card?]       // 4 slots (one per suit); nil = empty

    // MARK: - Game metadata

    let seed:     Int64
    let drawMode: String   // "draw1" | "draw3"
    var drawCount: Int { drawMode == "draw1" ? 1 : 3 }

    var turns:     String = ""    // comma-separated move tokens
    var moveCount: Int    = 0
    var isNew:     Bool   = true  // true until first move

    // MARK: - History (undo)

    private(set) var history: [GameState] = []  // snapshots for undo

    // MARK: - Init from seed

    init(seed: Int64, drawMode: String) {
        self.seed     = seed
        self.drawMode = drawMode

        let cards = SeededShuffle.deal(seed: seed)
        // Deal tableau: column i gets (i+1) cards, last is face up
        var idx = 0
        var cols = [[Card]]()
        for col in 0..<7 {
            var column = Array(cards[idx ..< idx + col + 1])
            for j in 0..<column.count { column[j].isFaceUp = (j == column.count - 1) }
            cols.append(column)
            idx += col + 1
        }
        self.tableau    = cols
        // Stock top = cards[idx] (first card drawn), matching Java's poll() order.
        // Swift uses removeLast() for O(1) pops, so the array is stored reversed:
        // stock.last is drawn first, stock.first is drawn last.
        self.stock      = Array(cards[idx...].reversed()).map { var c = $0; c.isFaceUp = false; return c }
        self.waste      = []
        self.foundation = [nil, nil, nil, nil]
    }

    // MARK: - Win detection

    var isWon: Bool {
        foundation.compactMap { $0 }.filter { $0.rank == .king }.count == 4
    }

    // MARK: - Waste top

    var wasteTop: Card? { waste.last }

    // MARK: - Move: Draw

    mutating func draw() {
        saveHistory()
        if stock.isEmpty {
            // Flip waste back to stock
            stock = waste.reversed().map { var c = $0; c.isFaceUp = false; return c }
            waste = []
        } else {
            let count = min(drawCount, stock.count)
            for _ in 0..<count {
                var card = stock.removeLast()
                card.isFaceUp = true
                waste.append(card)
            }
        }
        appendTurn("draw")
    }

    // MARK: - Move: Waste → Tableau

    @discardableResult
    mutating func moveWasteToTableau(col: Int) -> Bool {
        guard let card = waste.last,
              canPlaceOnTableau(card, column: col) else { return false }
        saveHistory()
        waste.removeLast()
        tableau[col].append(card)
        appendTurn("wt:\(col)")
        return true
    }

    // MARK: - Move: Waste → Foundation

    @discardableResult
    mutating func moveWasteToFoundation() -> Bool {
        guard let card = waste.last,
              let slot = foundationSlot(for: card) else { return false }
        saveHistory()
        waste.removeLast()
        foundation[slot] = card
        appendTurn("wf")
        return true
    }

    // MARK: - Move: Tableau → Foundation

    @discardableResult
    mutating func moveTableauToFoundation(col: Int) -> Bool {
        guard let card = tableau[col].last,
              let slot = foundationSlot(for: card) else { return false }
        saveHistory()
        tableau[col].removeLast()
        flipTopIfNeeded(col: col)
        foundation[slot] = card
        appendTurn("tf:\(col)")
        return true
    }

    // MARK: - Move: Tableau → Tableau

    @discardableResult
    mutating func moveTableau(fromCol: Int, fromIdx: Int, toCol: Int) -> Bool {
        guard fromCol != toCol,
              fromIdx < tableau[fromCol].count else { return false }
        let stack = Array(tableau[fromCol][fromIdx...])
        guard let top = stack.first,
              canPlaceOnTableau(top, column: toCol) else { return false }
        saveHistory()
        tableau[fromCol].removeSubrange(fromIdx...)
        flipTopIfNeeded(col: fromCol)
        tableau[toCol].append(contentsOf: stack)
        appendTurn("tt:\(fromCol):\(fromIdx):\(toCol)")
        return true
    }

    // MARK: - Move: Foundation → Tableau

    @discardableResult
    mutating func moveFoundationToTableau(foundationIdx: Int, toCol: Int) -> Bool {
        guard let card = foundation[foundationIdx],
              canPlaceOnTableau(card, column: toCol) else { return false }
        saveHistory()
        foundation[foundationIdx] = previousFoundationCard(suitIndex: foundationIdx)
        tableau[toCol].append(card)
        appendTurn("ft:\(foundationIdx):\(toCol)")
        return true
    }

    // MARK: - Replay

    /// Replay a comma-separated turns string produced by the Python/Java solver.
    ///
    /// Mirrors `GameState.replay()` in the Spring Boot backend (GameState.java)
    /// so that any valid server-side turns string also validates here.
    ///
    /// - Returns: `true` if all tokens applied without error, `false` on the
    ///   first invalid move (with `errorMessage` set on the returned copy).
    @discardableResult
    mutating func replay(turns: String) -> Bool {
        guard !turns.isEmpty else { return false }
        for token in turns.split(separator: ",").map(String.init) {
            let parts = token.split(separator: ":").map(String.init)
            guard !parts.isEmpty else { continue }
            var ok = false
            switch parts[0] {
            case "draw":
                draw()
                ok = true
            case "wf":
                ok = moveWasteToFoundation()
            case "wt":
                if parts.count >= 2, let col = Int(parts[1]) {
                    ok = moveWasteToTableau(col: col)
                }
            case "tf":
                if parts.count >= 2, let col = Int(parts[1]) {
                    ok = moveTableauToFoundation(col: col)
                }
            case "tt":
                if parts.count >= 4,
                   let from = Int(parts[1]),
                   let fi   = Int(parts[2]),
                   let to   = Int(parts[3]) {
                    ok = moveTableau(fromCol: from, fromIdx: fi, toCol: to)
                }
            case "ft":
                if parts.count >= 3,
                   let fi  = Int(parts[1]),
                   let col = Int(parts[2]) {
                    ok = moveFoundationToTableau(foundationIdx: fi, toCol: col)
                }
            default:
                break
            }
            if !ok { return false }
        }
        return true
    }

    // MARK: - Undo

    mutating func undo() {
        guard !history.isEmpty else { return }
        let prev = history.removeLast()
        // Restore board without clobbering the history array itself
        stock       = prev.stock
        waste       = prev.waste
        tableau     = prev.tableau
        foundation  = prev.foundation
        turns       = prev.turns
        moveCount   = prev.moveCount
    }

    // MARK: - Auto-complete eligibility

    /// True when all remaining cards are face up (no hidden tableau cards).
    var canAutoComplete: Bool {
        !isWon && stock.isEmpty && waste.isEmpty &&
        tableau.allSatisfy { col in col.allSatisfy { $0.isFaceUp } }
    }

    // MARK: - Private helpers

    private mutating func saveHistory() {
        var snapshot = self
        snapshot.history = []   // don't nest history snapshots
        history.append(snapshot)
        if history.count > 100 { history.removeFirst() }  // cap memory
        isNew = false
    }

    private mutating func appendTurn(_ token: String) {
        turns = turns.isEmpty ? token : "\(turns),\(token)"
        moveCount += 1
    }

    private mutating func flipTopIfNeeded(col: Int) {
        guard !tableau[col].isEmpty,
              !tableau[col][tableau[col].count - 1].isFaceUp else { return }
        tableau[col][tableau[col].count - 1].isFaceUp = true
    }

    private func canPlaceOnTableau(_ card: Card, column: Int) -> Bool {
        if let top = tableau[column].last {
            return card.canStackOnTableau(top)
        }
        return card.rank == .king  // empty column accepts only Kings
    }

    private func foundationSlot(for card: Card) -> Int? {
        let slot = card.suit.rawValue
        if let top = foundation[slot] {
            return card.rank.rawValue == top.rank.rawValue + 1 ? slot : nil
        }
        return card.rank == .ace ? slot : nil
    }

    private func previousFoundationCard(suitIndex: Int) -> Card? {
        guard let current = foundation[suitIndex],
              current.rank != .ace else { return nil }
        let prevId = current.id - 1
        return Card(id: prevId, isFaceUp: true)
    }
}
